fabric-rendering-v1/src/client/java/net/fabricmc/fabric/api/client/rendering/v1/world/WorldRenderEvents.java

```diff
@@ -0,0 +1,256 @@
+	 *
+	 * <p>Note that 3rd-party renderers may combine these passes or otherwise alter the
+	 * rendering pipeline for sake of performance or features. This can break direct writes to the
+	 * framebuffer.  Use this event for cases that cannot be satisfied by FabricBakedModel,
+	 * BlockEntityRenderer or other existing abstraction. If at all possible, use an existing terrain
+	 * RenderLayer instead of outputting to the framebuffer directly with GL calls.
+	 *
+	 * <p>The consumer is responsible for setup and tear down of GL state appropriate for the intended output.
+	 *
+	 * <p>Because solid and cutout quads are depth-tested, order of output does not matter except to improve
+	 * culling performance, which should not be significant after primary terrain rendering. This means
+	 * mods that currently hook calls to individual render layers can simply execute them all at once when
+	 * the event is called.
+	 *
+	 * <p>This event fires before entities and block entities are rendered and may be useful to prepare them.
+	 * However, you should not access any data outside the provided render states. If more data is needed,
+	 * extract them during {@link #END_EXTRACTION}.
+	 */
+	public static final Event<BeforeEntities> BEFORE_ENTITIES = EventFactory.createArrayBacked(BeforeEntities.class, callbacks -> context -> {
+		for (final BeforeEntities callback : callbacks) {
+			callback.beforeEntities(context);
+		}
+	});
+
+	/**
+	 * Called after entities and block entities are rendered and drawn to the main frame buffer target.
+	 */
+	public static final Event<AfterEntities> AFTER_ENTITIES = EventFactory.createArrayBacked(AfterEntities.class, callbacks -> context -> {
+		for (final AfterEntities callback : callbacks) {
+			callback.afterEntities(context);
+		}
+	});
+
+	/**
+	 * Called before vanilla debug renderers are output to the framebuffer.
+	 * This happens very soon after entities, block breaking and most other
+	 * non-translucent renders but before translucency is drawn.
+	 *
+	 * <p>The OpenGL render state view matrix will be transformed to match the camera view
+	 * before the event is called.
+	 *
+	 * <p>Use to drawn lines, overlays and other content similar to vanilla
+	 * debug renders.
+	 */
+	public static final Event<DebugRender> BEFORE_DEBUG_RENDER = EventFactory.createArrayBacked(DebugRender.class, callbacks -> context -> {
+		for (final DebugRender callback : callbacks) {
+			callback.beforeDebugRender(context);
+		}
+	});
+
+	/**
+	 * Called after entity, terrain, and particle translucent layers have been
+	 * drawn to the framebuffer but before translucency combine has happened
+	 * in fabulous mode.
+	 *
+	 * <p>Use for drawing overlays or other effects on top of those targets
+	 * (or the main target when fabulous isn't active) before clouds and weather
+	 * are drawn.
+	 */
+	public static final Event<BeforeTranslucent> BEFORE_TRANSLUCENT = EventFactory.createArrayBacked(BeforeTranslucent.class, callbacks -> context -> {
```

Comment:
The documentation for this event is incorrect:
- It is called before translucent terrain, not after.
- I do not remember when particles are drawn. They may be drawn outside of the main pass, in which case this event is not called after particle translucent layers are drawn.
- It should not be used for drawing effects on top of fabulous targets. `END_MAIN` should be used for that one.